### ISSUE TITLE
How to profile CUDA kernels in PyTorch

### ISSUE THREAD

**Issue by solangii** [solangii/gpu-practice#1]
## Contents
1. Pytorch Autograd Profiler
2. Pytorch profiler
3. Intergration CUDA kernels in PyTorch
4. Triton
5. Optimization and Profiling with Nsight Compute

https://www.youtube.com/watch?v=LuhJEEJQgUM&t

**Comment by solangii** [comment]
## Profiling PyTorch Square with Autograd Profiler

### Torch Autograd Profiler
- CPU, GPU에서 kernel execution time, # of calls, dependencies에 대한 인사이트 제공해줌
- CUDA는 비동기식, 따라서 profiling tool이 필요하다. 
  - `Event` : 동기화 마커. 진행상황을 모니터링하고 타이밍을 정확하게 측정하기 위함. - [document](https://pytorch.org/docs/stable/generated/torch.cuda.Event.html#torch.cuda.Event)
  -  `profiler`: CPU/GPU 모두에서 모델 내부 operator의 cost 검사 - [document](https://pytorch.org/docs/stable/autograd.html#profiler
)

### Profile Square Tensor
- `torch.square`, python multiplication, python power operation 비교
- ![Image](https://github.com/user-attachments/assets/2f6f3c9a-9eb8-469f-8b66-28b8ad885120)
- 각 함수를 실행했을 때 호출되는 C++ function들 확인 가능.
- `torch.squre`의 경우`aten::squre` C++ 실행함. `aten::squre`는 `aten:pow` w/ `2`를 호출 - [code](https://github.com/pytorch/pytorch/blob/b8b04b26fbf160874f7f1a9db61e49801fd4fcbe/aten/src/ATen/native/UnaryOps.cpp#L771
)
- `a*a`는 `aten::mul` 호출
- `a**2`는 `aten::pow` 호출 (`torch.squre`에서도 square -> pow호출. 따라서 time cost도 유사함)

**Comment by solangii** [comment]
## Pytorch Profiler
- [document](https://pytorch.org/docs/stable/profiler.html)
- visual profiler w/ [Chrome trace](chrome://tracing/). (JSON파일을 여기에 drop하면 볼 수 있음.)
- memory copies, kernel launches, flow event에 대한 정보 제공. 커널 성능 혹은 개선 방법에 대한 정보는 제공하지 않음

### Profiling `torch.square()`
- send tensor to GPU
- Compute square of the tensor

#### Default usage
https://github.com/solangii/gpu-practice/blob/76804cad50bc454f016b9d6b854a63ffa317bd2d/profile/pt_profiler-default.py#L5-L10

#### Profiling Results
![Image](https://github.com/user-attachments/assets/d015cce7-5f39-454c-baf8-dc4a7b363227)

#### custom usage - w/ schedule
https://github.com/solangii/gpu-practice/blob/76804cad50bc454f016b9d6b854a63ffa317bd2d/profile/pt_profiler-custom.py#L5-L42

#### Profiling Results
![Image](https://github.com/user-attachments/assets/ccf8c0d7-4cf8-4036-8ffc-3b22644bc962)
![Image](https://github.com/user-attachments/assets/82d589f7-dc1a-48a2-91e6-aa1baae4caac)

- 나는 왜 HtoD copy가 안뜨지 ㅠㅠ.. profilerStop*이전 단계가 내꺼에선 안뜨는 문제 ㅠ

**Comment by solangii** [comment]
## Integrating CUDA Kernels in PyTorch
- CUDA is written by C/C++
- PyBind: Create Python bindings for C++ code - [document](https://pybind11.readthedocs.io/en/stable/)
- torch.utils.cpp_extension.load_inline  - [document](https://pytorch.org/docs/stable/cpp_extension.html#torch.utils.cpp_extension.load_inline)
  -  C++ souurce code를 argument로 전달하면, 자동으로 C++ with pybind /CUDA with headers /`build.ninja` for compiling.. 생성해줌.

### Hello World Example
``` 
$ python3 load_inline_hello_world.py                                                                                                                                                          ✔  
Emitting ninja build file ./load_inline_hello_world_cuda/build.ninja...
Building extension module my_module...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
[1/2] c++ -MMD -MF main.o.d -DTORCH_EXTENSION_NAME=my_module -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -isystem /home/solang/.local/lib/python3.10/site-packages/torch/include -isystem /home/solang/.local/lib/python3.10/site-packages/torch/include/torch/csrc/api/include -isystem /home/solang/.local/lib/python3.10/site-packages/torch/include/TH -isystem /home/solang/.local/lib/python3.10/site-packages/torch/include/THC -isystem /usr/include/python3.10 -D_GLIBCXX_USE_CXX11_ABI=0 -fPIC -std=c++17 -c /home/solang/gpu-practice/profile/load_inline_hello_world_cuda/main.cpp -o main.o
[2/2] c++ main.o -shared -L/home/solang/.local/lib/python3.10/site-packages/torch/lib -lc10 -ltorch_cpu -ltorch -ltorch_python -o my_module.so
Loading extension module my_module...
Hello World!
```
- `load_inline_hello_world.py` 실행시 load_inline에 의해 자동으로 빌드 된 것을 볼 수 있음

### Custom CUDA Kernel for Square Operation
- with CUDA Kernel
- CUDA 커널의 입력으로 pytorch 텐서를 준비하는 Wrapper function
-  이를 통해 'square_matrix' 함수를 네이티브 파이토치 함수처럼 사용할 수 이씀.
```python
# Load the CUDA kernel as a PyTorch extension
square_matrix_extension = load_inline(
    name='square_matrix_extension',
    cpp_sources=cpp_source, 
    cuda_sources=cuda_source,
    functions=['square_matrix'],
    with_cuda=True,
    extra_cuda_cflags=["-O2"], # compiler optimization
    build_directory='./load_inline_cuda',
    # extra_cuda_cflags=['--expt-relaxed-constexpr']
)
```
- 위 코드 실행시 cuda code 구할 수 있음. - 정확히 무슨 용도인지 아직 이해 못했음..
- 대체제로 Numba가 있음. - [website](https://numba.pydata.org/)
  - Write CUDA kernels directly in Python.
  - Easier syntax compared to C++, but may have performance limitations.

**Comment by solangii** [comment]
## Triton
- [document](https://triton-lang.org/main/index.html)
- Python-based domain-specific language (DSL) for GPU programming.
- Block-based programming language
- Does not generate CUDA. 
  - Generates a PTX kernel (CUDA assembly)
- Has a cache called .triton
- The most important optimization for machine learning code is often fusions
  - Include more operations in a single kernel

### Square operation using Triton
- Triton을 사용하여 GPU에서 행렬의 각 요소를 제곱(x^2)하는 커널을 작성(`square_kernel`)하고, 이를 PyTorch의 `torch.square` 연산과 성능 비교하는 벤치마크 테스트 수행

- Triton 커널이란
  - 각 행(row) 별로 독립적인 연산을 수행 → 행렬의 각 요소를 병렬 처리
  - Triton의 @triton.jit 데코레이터 사용 → CUDA 커널을 JIT(just-in-time) 컴파일하여 최적화

#### Kernel
```python
@triton.jit
def square_kernel(output_ptr, input_ptr, input_row_stride, output_row_stride, n_cols, BLOCK_SIZE: tl.constexpr):
    row_idx = tl.program_id(0)  # 현재 실행 중인 행(row)의 인덱스
    row_start_ptr = input_ptr + row_idx * input_row_stride  # 현재 행의 시작 위치
    col_offsets = tl.arange(0, BLOCK_SIZE)  # 열(column) 오프셋
    input_ptrs = row_start_ptr + col_offsets  # 각 열 위치의 메모리 주소
    row = tl.load(input_ptrs, mask=col_offsets < n_cols, other=-float('inf'))  # 데이터 로드

    square_output = row * row  # 각 요소를 제곱

    output_row_start_ptr = output_ptr + row_idx * output_row_stride  # 출력 행 시작 위치
    output_ptrs = output_row_start_ptr + col_offsets  # 출력 위치
    tl.store(output_ptrs, square_output, mask=col_offsets < n_cols)  # 결과 저장
```
- 각 행(row)마다 하나의 GPU 프로그램(instance)이 실행됨.
- `tl.store(output_ptrs, square_output, mask=...)` → 결과를 GPU 메모리에 저장.

#### Triton 커널을 파이썬에서 호출하는 함수
```python
def square(x):
    n_rows, n_cols = x.shape
    BLOCK_SIZE = triton.next_power_of_2(n_cols)  # 가장 가까운 2의 제곱수를 블록 크기로 설정
    num_warps = 4 if BLOCK_SIZE < 2048 else (8 if BLOCK_SIZE < 4096 else 16)  # 워프 수 설정
    y = torch.empty_like(x)  # 결과를 저장할 PyTorch 텐서 생성

    square_kernel[(n_rows,)](  # Triton 커널 실행
        y, x, x.stride(0), y.stride(0), n_cols,
        num_warps=num_warps,
        BLOCK_SIZE=BLOCK_SIZE,
    )
    return y
```
- `triton.next_power_of_2(n_cols)` → 열 개수(n_cols)보다 크거나 같은 가장 가까운 2의 거듭제곱을 찾음.
- `num_warps`는 행렬 크기에 따라 다르게 설정 (최적화 목적).
- `square_kernel[(n_rows, )](...)` → Triton 커널을 실행하여 행렬의 제곱 연산을 수행.


#### Benchmark
- 다양한 크기의 행렬에 대해 성능 측정 (N 값 변화)
- Triton 커널 vs. PyTorch 기본 연산 vs. PyTorch Compile 연산 비교
![Image](https://github.com/user-attachments/assets/40cc3bcb-cfc8-4e54-886d-addfda5b5e41)
- 결과 왜 차이가 없을까.. 내 gpu가 구려서 그런가..


### Triton debugger
- Use `triton.jit(interpret=True)`
- 혹은 환경변수로 `TRITON_INTERPRET=1` 사용.
- 대부분이 WrappedTensor이므로 `var_name.tensor`로변수 inspecting가능
- 주의사항, Can’t use print statements! (global memory에 변수 저장하고 read하기 땜)

### Exploration of generated Square Kernel PTX (cuda 어셈블리)
- Element-wise matrix square
- Triton is leveraging 8 registers at a time - 그닥 효율적이진 않은?
- Store global: writing the variables back to global memory
  - Can see what are the actual registers being used directly
- Recommendation: Use ChatGPT (or other) to annotate the generated PTX assembly code

### Code Example: Auto-generate a triton kernel using torch.compile()

**Comment by solangii** [comment]
## Optimization & Profiling with Nsight Compute
- NVIDIA NSight Compute Profiler
- [download](https://developer.nvidia.com/tools-overview/nsight-compute/get-started)
- if remote machine의 gpu로 해야한다면? [참고](https://tspeterkim.github.io/posts/nsight-setup-on-ec2)

**Comment by solangii** [comment]
## Next Step?

- CUDA 기본 문법, 연산방식(thread/block/grid 구조)에 대한 이해 필요
- Triton / torch연산 차이 없는 이유? 내 머신에서만 그런 것인지?
-
